google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/ConnectionImpl.java

```diff
@@ -995,33 +995,63 @@ GetQueryResultsResponse getQueryResultsFirstPage(JobId jobId) {
-              () ->
-                  bigQueryRpc.getQueryResultsWithRowLimit(
-                      completeJobId.getProject(),
-                      completeJobId.getJob(),
-                      completeJobId.getLocation(),
-                      connectionSettings.getMaxResultPerPage()),
-              bigQueryOptions.getRetrySettings(),
-              BigQueryBaseService.BIGQUERY_EXCEPTION_HANDLER,
-              bigQueryOptions.getClock(),
-              retryConfig);
 
-      if (results.getErrors() != null) {
-        List<BigQueryError> bigQueryErrors =
-            results.getErrors().stream()
-                .map(BigQueryError.FROM_PB_FUNCTION)
-                .collect(Collectors.toList());
-        // Throwing BigQueryException since there may be no JobId and we want to stay consistent
-        // with the case where there there is a HTTP error
-        throw new BigQueryException(bigQueryErrors);
+    // Implementing logic to poll the Job's status using getQueryResults as
+    // we do not get rows, rows count and schema unless the job is complete
+    // Ref: b/241134681
+    // This logic will wait for approx (poolingIntervalMs + 10 seconds which is the default timeout
+    // for getQueryResults) per iteration of the loop
+    long startTimeMs = System.currentTimeMillis();
+    long totalTimeOutMs = 18 * 60 * 60 * 1000; // 18 hours, which is the max timeout for the job
+    long poolingIntervalMs = 60000; // 1 min
+    GetQueryResultsResponse results = null;
+
+    while ((System.currentTimeMillis() - startTimeMs) <= totalTimeOutMs) {
+      try {
+        results =
+            BigQueryRetryHelper.runWithRetries(
+                () ->
+                    bigQueryRpc.getQueryResultsWithRowLimit(
+                        completeJobId.getProject(),
+                        completeJobId.getJob(),
+                        completeJobId.getLocation(),
+                        connectionSettings.getMaxResultPerPage()),
+                bigQueryOptions.getRetrySettings(),
+                BigQueryBaseService.BIGQUERY_EXCEPTION_HANDLER,
+                bigQueryOptions.getClock(),
+                retryConfig);
+
+        if (results.getErrors() != null) {
+          List<BigQueryError> bigQueryErrors =
+              results.getErrors().stream()
+                  .map(BigQueryError.FROM_PB_FUNCTION)
+                  .collect(Collectors.toList());
+          // Throwing BigQueryException since there may be no JobId and we want to stay consistent
+          // with the case where there  is a HTTP error
+          throw new BigQueryException(bigQueryErrors);
+        }
+      } catch (BigQueryRetryHelper.BigQueryRetryHelperException e) {
+        throw BigQueryException.translateAndThrow(e);
+      }
+
+      if (results.getJobComplete()) {
+        break; // The job is complete
+
+      } else { // wait for the defined poolingIntervalMs and the loop will retry
+        try {
+          Thread.sleep(poolingIntervalMs);
```

Comment:
The behavior here seems wrong.  getQueryResults will poll from the server side for up to 10 seconds, and then you sleep for 60 in the client?  Effective latency for an 11 sec query becomes more than a minute.

You should not wait on the client side at all, just re-issue a subsequent getQueryResults call.  Allow the poll and wait to happen within BQ.

Reply:
Done, updated to logic to use RCP based poll and wait